numpy/core/src/multiarray/descriptor.c

```diff
@@ -1892,18 +1892,28 @@ arraydescr_protocol_typestr_get(PyArray_Descr *self)
     else {
         ret = PyUnicode_FromFormat("%c%c%d", endian, basic_, size);
     }
+    if (ret == NULL) {
+        return NULL;
+    }
+
     if (PyDataType_ISDATETIME(self)) {
         PyArray_DatetimeMetaData *meta;
-
         meta = get_datetime_metadata_from_dtype(self);
         if (meta == NULL) {
             Py_DECREF(ret);
             return NULL;
         }
+        PyObject *umeta = metastr_to_unicode(meta, 0);
+        if (umeta == NULL) {
+            Py_DECREF(ret);
+            return NULL;
+        }
 
-        ret = append_metastr_to_string(meta, 0, ret);
+        PyObject *tmp = ret;
+        ret = PyUnicode_Concat(umeta, ret);
+        Py_DECREF(tmp);
```

Comment:
Could if you wanted use
```suggestion
        Py_SETREF(ret, PyUnicode_Concat(umeta, ret));
```